tests/test_boltzmann_machine.py

```diff
@@ -356,6 +356,46 @@ def test_objective(self):
             self.assertEqual(-1, grbm.quasi_objective(s1, s2).item())
             self.assertEqual(-1, grbm.quasi_objective(s1, s3))
 
+    def test_quasi_objective_gradient_hidden_units(self):
+        grbm = GRBM([1, 2, 3],
+                    [(1, 2), (1, 3), (2, 3)],
+                    [1],
+                    {1: 0.2, 2: 0.2, 3: 0.3},
+                    {(1, 2): 0.2, (1, 3): 0.3, (2, 3): 0.6})
+        # Note : In the digram bellow linear biases are shown using  <>
+        #        quadratic biases using ()
+        #                 (0.2)
+        # Model:  v1 <0.2> ----- v2  <0.2>
+        #           \           /
+        #     (0.3)  \         / (0.6)
+        #             \       /
+        #                v3 <0.3>
+        s_observed = torch.tensor([[1.0, -1.0]])
+        s_model = torch.tensor([[1.0, -1.0, 1.0]])
+        quasi = grbm.quasi_objective(s_observed, s_model, "exact-disc")
+        quasi.backward()
+        # Compute gradients manually
+        # Compute unnormalized density
+        #                            h1v1 + h2v2 + h3v3 + J23v2v3 + J12v1v2 + J13v1v3
+        q_plus = torch.exp(-torch.tensor(0.2 + 0.2 - 0.3 - 0.6 + 0.2 - 0.3))
+        q_minus = torch.exp(-torch.tensor(-0.2 - 0.2 + 0.3 - 0.6 + 0.2 - 0.3))
```

Comment:
looks like this is not sorted  according to `h1v1 + h2v2 + h3v3 + J23v2v3 + J12v1v2 + J13v1v3`:
I think it should be
 q_plus = torch.exp(-torch.tensor(0.2  +  0.2  - 0.3  - 0.6 + 0.2 - 0.3))
 q_minus = torch.exp(-torch.tensor(-0.2 + 0.2 - 0.3 - 0.6 - 0.2 + 0.3))
 this won't change the results

Reply:
great catch, thanks!